File: leetcode/editor/en/[1063]Number of Valid Subarrays.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number[]} nums
 * @return {number}
 */
var validSubarrays = function (nums) {
    let result = 0;
    let stack = [];

    for (let i = 0; i < nums.length; i++) {
        while (stack.length > 0 && nums[i] < nums[stack[stack.length - 1]]) {
            result += (i - stack[stack.length - 1]);
            stack.pop();
        }
        stack.push(i);
    }

    while ((stack.length > 0)) {
        result += (nums.length - stack[stack.length - 1]);
        stack.pop();
    }

    return result;
};
//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/en/[1063]Number of Valid Subarrays.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int validSubarrays(int[] nums) {
        int result = 0;
        Deque<Integer> dq = new ArrayDeque<>();

        for (int i = 0; i < nums.length; i++) {
            while (!dq.isEmpty() && nums[i] < nums[dq.peek()]) {
                result += (i - dq.peek());
                dq.pop();
            }
            dq.push(i);
        }

        while (!dq.isEmpty()) {
            result += (nums.length - dq.peek());
            dq.pop();
        }

        return result;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
